### PR TITLE
Feat: Add Union Find Algorithm, Test: Add test for Union Find Algorithm

### PR DESCRIPTION
**Pull request by MugdhaBehere**
Added the Union Find(Dynamic Connectivity) Algorithm and the test file for that.

- [x] Added description of change
- [x] Added file name matches File name guidelines.
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic commit guidelines
- [x]  Search previous suggestions before making a new one, as yours may be a duplicate
- [x]  I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Comment by MugdhaBehere** [comment]
If this pull request is mergeable, I request the reviewers or code owners to merge it by labeling it as hacktoberfest-accepted.

File: graph/unionfind.go
Comment (tjgurwara99) [inline graph/unionfind.go:13]:
```suggestion
// Union Find Algorithm or Dynamic Connectivity algorithm, often implemented with the help of the union find data structure,
// is used to efficiently maintain connected components in a graph that undergoes dynamic changes,
// such as edges being added or removed over time
// Worst Case Time Complexity: The time complexity of find operation is nearly constant or O(α(n)), where where α(n) is the inverse Ackermann function
// practically, this is a very slowly growing function making the time complexity for find operation nearly constant.
// The time complexity of the union operation is also nearly constant or O(α(n))
// Worst Case Space Complexity: O(n), where n is the number of nodes or element in the structure
// Reference: https://www.scaler.com/topics/data-structures/disjoint-set/
// Author: Mugdha Behere[https://github.com/MugdhaBehere]
// see: unionfind.go, unionfind_test.go

package graph

import "fmt"
```

**Review by tjgurwara99 — COMMENTED** [review]
I believe we already [have the](https://github.com/TheAlgorithms/Go/blob/5206ad3945b2a8b58bb0a89ba78962401925dca6/graph/kruskal.go#L29) `DisjointSetUnion` in this repository - I can't quire remember if that is exactly the same as Union Find but I do remember them being the same. @raklaptudirm @siriak any opinions on this?

File: graph/unionfind.go
Comment (MugdhaBehere) [reply graph/unionfind.go]: These changes have been made.

**Comment by MugdhaBehere** [comment]
Yes, a similar or same algorithm has been applied to implement Kruskal's Algorithm. This is a separate implementation if required, though. That is an example where the union-find algorithm has been applied. I did not find a separate implementation of the Union-find algorithm in this repository. If you find the contribution considerable and if this algorithm is required and the pull request is mergeable, I request the code owners or reviewers to merge it with the hacktoberfest-accepted label.

**Comment by siriak** [comment]
So, disjoint set union (DSU) and union-find (UF) are the same. Could you remove DSU from Kruskals algorithm and use your UF there? Or merge DSU and UF so that they are the same and separate from Kruskals algorithm?

**Comment by siriak** [comment]
Yes, please use it in Kruskal's algo. Let's do it in this PR so that we don't have 2 implementations at once

**Comment by MugdhaBehere** [comment]
Kruskal's algorithm has been updated as required by you. I request the reviewers to go through the pull request and if it is mergeable, merge it by labeling it as hacktoberfest-accepted.

**Comment by MugdhaBehere** [comment]
@siriak @tjgurwara99 Sir, could you please go through the pull request and the changes I made? I have updated Kruskal's algorithm as required by you. And, I have also removed the errors I got in the unionfind_test.go file.

**Review by siriak — CHANGES_REQUESTED** [review]
Please reuse the implementation of UF in `graph/unionfind.go` by calling it in `graph/kruskal.go`. No need to implement it twice

**Comment by MugdhaBehere** [comment]
I have removed the redundant union-find implementation from Kruskal's algorithm. Please check.

**Comment by siriak** [comment]
Now the implementation looks good. Please remove `main` functions both in UF and Kruskals to make CI pass

**Review by siriak — COMMENTED** [review]
image
The code looks good, could you please fix formatting so that PR checks pass?

**Comment by MugdhaBehere** [comment]
I have tried to remove all errors. Sir, could you please check if there are any more errors/ problems with the code style?

File: graph/unionfind.go
Comment (tjgurwara99) [inline graph/unionfind.go:25]:
```suggestion
// Union Find Algorithm or Dynamic Connectivity algorithm, often implemented with the help 
//of the union find data structure,
// is used to efficiently maintain connected components in a graph that undergoes dynamic changes,
// such as edges being added or removed over time
// Worst Case Time Complexity: The time complexity of find operation is nearly constant or 
//O(α(n)), where where α(n) is the inverse Ackermann function
// practically, this is a very slowly growing function making the time complexity for find 
//operation nearly constant.
// The time complexity of the union operation is also nearly constant or O(α(n))
// Worst Case Space Complexity: O(n), where n is the number of nodes or element in the structure
// Reference: https://www.scaler.com/topics/data-structures/disjoint-set/
// Author: Mugdha Behere[https://github.com/MugdhaBehere]
// see: unionfind.go, unionfind_test.go
```

You don't need to add empty lines between comments.

File: graph/unionfind.go
Comment (tjgurwara99) [inline graph/unionfind.go:48]: Use the go doc styled commenting conventions - https://tip.golang.org/doc/comment

**Review by tjgurwara99 — COMMENTED** [review]
Please read the documentation conventions for Go - https://tip.golang.org/doc/comment and make the necessary changes.

File: graph/kruskal.go
Comment (tjgurwara99) [inline graph/kruskal.go:39]: You can just use `I` instead of casting it to `Vertex` then casting it to an `int`.

**Comment by tjgurwara99** [comment]
Please also format your code with `gofmt`

File: graph/unionfind.go
Comment (MugdhaBehere) [reply graph/unionfind.go:34]: This change has been made

**Comment by MugdhaBehere** [comment]
I had previously formatted the code with gofmt. According to the error message I could see now, I formatted it again, with gofmt -s, but there were no new changes to commit even after formatting it again with gofmt -s. Please check.

**Comment by MugdhaBehere** [comment]
I have tried to change the files as required by you. I request the reviewers to go through the changes. I have formatted the files with the help of -w and -s flags at the same time. For example, to format unionfind.go, I wrote the command,          
                    gofmt -w -s graph/unionfind.go
Similarly, I wrote commands for all files.

**Review by siriak — APPROVED** [review]
Looks good, thanks!

**Comment by MugdhaBehere** [comment]
I request the reviewers or code owners to review the changes for approval

**Comment by raklaptudirm** [comment]
@tjgurwara99 I am waiting for your review since you already reviewed this pr once.

**Review by tjgurwara99 — APPROVED** [review]
Looks alright to me - the tests are not blackbox tested. But not requesting the author to change this - might change it myself soon since I'm a bit concerned about its API but that's fine - I'll check that. This PR can be merged.

**Pull request merged**
